DuckDuckGo/AutofillLoginDetailsViewModel.swift

```
@@ -0,0 +1,182 @@
+        self.updateData(with: account)
+    }
+    
+    private func updateData(with account: SecureVaultModels.WebsiteAccount) {
+        self.username = account.username
+        self.address = account.domain
+        self.title = account.name
+        self.headerViewModel.updateData(with: account)
+        setupPassword(with: account)
+    }
+    
+    func toggleEditMode() {
+        withAnimation {
+            if viewMode == .edit {
+                viewMode = .view
+            } else {
+                viewMode = .edit
+            }
+        }
+    }
+    
+    func copyToPasteboard(_ action: PasteboardCopyAction) {
+        var itemName = ""
+        switch action {
+        case .username:
+            itemName = "Username"
+            UIPasteboard.general.string = username
+        case .password:
+            itemName = "Password"
+            UIPasteboard.general.string = password
+        case .address:
+            itemName = "Address"
+            UIPasteboard.general.string = address
+        }
+        
+        presentCopyConfirmation(message: "\(itemName) copied")
+    }
+    
+    private func presentCopyConfirmation(message: String) {
+        DispatchQueue.main.async {
+            ActionMessageView.present(message: message,
+                                      actionTitle: "",
+                                      onAction: {})
+        }
+    }
+    
+    private func setupPassword(with account: SecureVaultModels.WebsiteAccount) {
+        do {
+            if let accountID = account.id {
+                let vault = try SecureVaultFactory.default.makeVault(errorReporter: SecureVaultErrorReporter.shared)
+                
+                if let credential = try
+                    vault.websiteCredentialsFor(accountId: accountID) {
+                    self.password = String(data: credential.password, encoding: .utf8) ?? ""
+                }
+            }
+        } catch {
+            print("Can't retrieve password")
```

More to do here?

---

I've added a `Pixel.fire(pixel: .secureVaultError)` here CC/ @THISISDINOSAUR

---

Do you know what else will happen in these error cases? Do they fail gracefully?

---

@THISISDINOSAUR there was no error handling rather than firing a pixel on V1. Things won't crash, but on this specific case the password on the UI would be empty.

I wonder what could be done here, if securevault doesn't work for some reason the user can't take any action on it, should we display a generic error message? Fail silently (like currently)?

---

I think that's fine for now, I just wanted to check it. doesn't end up locking up the UI or anything like that